### PR TITLE
8321533: Clang build for Windows

### PR DESCRIPTION
**Pull request by djelinski**
I'd like to propose a new toolchain for Windows using the clang-cl compiler frontend.

Clang-cl is available as an optional feature in all Visual Studio editions, including the free-for-OSS-development community edition.

Clang-cl command line is mostly compatible with cl. However, clang-cl offers a distinct set of diagnostic messages and warnings, which can be used to improve code quality.

In order to use the clang compiler:
- install Visual Studio 2022
- install C++ Clang Compiler (https://learn.microsoft.com/en-us/cpp/build/clang-support-msbuild?view=msvc-170)
- configure `--with-toolchain-type=clcl`
- compile as usual.

Note: at this moment the code does not build; I plan to submit separate patches that fix the build problems separately. For the impatient, [this patch](https://github.com/djelinski/jdk/commit/d67a16244f4e6db8e6e8c59266bebd17827bc2a5) should be enough to compile the JDK, and [this one](https://github.com/djelinski/jdk/commit/1580e7cf54cca61d3ab58891619553994ea26b10) is needed to get tier1 to pass.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8321533](https://bugs.openjdk.org/browse/JDK-8321533): Clang build for Windows (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17019/head:pull/17019` \
`$ git checkout pull/17019`

Update a local copy of the PR: \
`$ git checkout pull/17019` \
`$ git pull https://git.openjdk.org/jdk.git pull/17019/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17019`

View PR using the GUI difftool: \
`$ git pr show -t 17019`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17019.diff">https://git.openjdk.org/jdk/pull/17019.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17019#issuecomment-1846012926)

### CODE REVIEW

File: make/autoconf/spec.gmk.in [hunk 1]
```diff
@@ -164,6 +164,11 @@ ifeq (@TOOLCHAIN_TYPE@, microsoft)
   # Visual Studio tools.
   export PATH := @TOOLCHAIN_PATH@:$(PATH)
 endif
+ifeq (@TOOLCHAIN_TYPE@, clcl)
```
Comment (erikj79) [inline make/autoconf/spec.gmk.in:167]: You can combine this with the ifeq above using an ifneq like this:

```
ifneq ($(filter @TOOLCHAIN_TYPE@, microsoft clcl), )
```

File: make/common/NativeCompilation.gmk [hunk 2]
```diff
@@ -447,7 +447,7 @@ define SetupCompileNativeFileBody
     $$($1_OBJ): $$($1_OBJ_DEPS) | $$($$($1_BASE)_BUILD_INFO)
 	$$(call LogInfo, Compiling $$($1_FILENAME) (for $$($$($1_BASE)_BASENAME)))
 	$$(call MakeDir, $$(@D))
-        ifneq ($(TOOLCHAIN_TYPE), microsoft)
+        ifeq ($(findstring $(TOOLCHAIN_TYPE), microsoft clcl), )
```
Comment (erikj79) [inline make/common/NativeCompilation.gmk:450]: I know `findstring` is used frequently for this construct in the makefile, but I prefer `filter` in cases like this as it only matches complete words.

**Review by erikj79 — COMMENTED** [review]
I only commented on minor style issues. I would like to hear from others what they think about the validity of this approach.

Does it work with current devkits, or are they missing something? How is build performance relative to regular `cl`?

**Comment by TheShermanTanker** [comment]
Wow, you're really leaving me in the dust here :P

Is it possible to fold clang-cl into the microsoft toolchain instead, and just pass clang-cl as the compiler instead of cl?

**Comment by djelinski** [comment]
Thanks for the suggestions, I'll apply them shortly.

The current devkits do not contain the required clang frontend (clang-cl.exe, llvm-lib.exe and lld-link.exe).

I compared `make hotspot` performance, got the following results:
- default toolchain (msvc 19.35): 6 min 35 sec
- this PR (clang 15.0.1): 7 min 41 sec
- this PR with compiler warnings disabled: 6 min 47 sec

so it's going to be slightly slower than MSVC once we have all the warning suppressions in place.

**Comment by djelinski** [comment]
> Is it possible to fold clang-cl into the microsoft toolchain instead, and just pass clang-cl as the compiler instead of cl?

I suppose it's possible; clang-cl output is supposed to be compatible with link & lib, and most command line options are similar.

The thing is, when it comes to warning control, only W0-W4 are somewhat similar between clang-cl and microsoft; other than that, clang-cl accepts the same options as the regular clang. Folding clang-cl into microsoft would leave us with no way to suppress warnings.

**Comment by TheShermanTanker** [comment]
> > Is it possible to fold clang-cl into the microsoft toolchain instead, and just pass clang-cl as the compiler instead of cl?
> 
> I suppose it's possible; clang-cl output is supposed to be compatible with link & lib, and most command line options are similar.
> 
> The thing is, when it comes to warning control, only W0-W4 are somewhat similar between clang-cl and microsoft; other than that, clang-cl accepts the same options as the regular clang. Folding clang-cl into microsoft would leave us with no way to suppress warnings.

My main concern is that this occludes clang proper, if and when that becomes a thing. If so, the clcl options might become a little redundant when compared with the existing clang option. Maybe there can be some specialized code to detect if microsoft is really clang-cl in disguise, or perhaps this could use the clang option that we already have in the build system, and then have --with-toolchain-type=clang mean either clang-cl or clang proper when the platform is Windows

**Comment by djelinski** [comment]
Well I can't tell the future, but at this point in time, implementing support for clang-cl requires much less effort than implementing  support for the gcc clang frontend. The end result is equivalent - in both cases we use the same compiler.

**Comment by magicus** [comment]
This is interesting, but we must be careful to not rush things here. We're basically creating a chimera, mixing assumptions about "windows", "microsoft toolchain" and "clang".

Do I understand you correctly that clang-cl uses "mostly" the same arguments as cl, but uses the clang warnings system, e.g. -Wno-foo-loops instead of C1234?

That means we need to separate the warning option handling from all the other options. You say something about not disabling warnings adds a minute to the build time. I assume that means that even with your linked patches, the build will spew out warnings.

This is not something we can fix "afterwards"; we need to design the integration of clang-cl in a way that we can (hopefully) reuse existing clang warning disables.

**Comment by magicus** [comment]
And, oh, I must say that I think "clcl" as name for the toolchain is really bad. :-(

I think `microsoft-clang` would be the best and clearest name. Possibly, we should rename `microsoft` to `microsoft-cl`, and use `microsoft` as an user-facing alias. At least if the rest of the toolchain is 100% the same as in the current `microsoft` toolchain.

Doing that will present both a risk and an opportunity: if we are checking the toolchain using "findstring" for "microsoft", but we really only want to match `microsoft-cl`, then it will break, so any such instances must be discovered and checked. On the other hand, if we do want to treat `microsoft-cl` and `microsoft-clang` the same, then a "findstring" instead of a "filter" will actually do the trick.

**Comment by magicus** [comment]
> >    And, oh, I must say that I think "clcl" as name for the toolchain is really bad. :-(
>
> It was not my first choice. I tried clang-cl first; the scripts didn't like the dash. Then I tried clangcl, and some of the changes found their way into the MacOS build. findstring is tricky to get right.

Then we need to fix those places. As Erik has commented, `findstring` is usually the wrong function to use when what you really mean is `filter`.

**Comment by magicus** [comment]
> >    we need to design the integration of clang-cl in a way that we can (hopefully) reuse existing clang warning disables.
> 
> I'd love that. Got any preferences on how that should be handled?

Not straight out of my hat, no. I wonder what would happen as a first step if we started looking at the DISABLE_WARNINGS_clang fields, by just using a hack. I'm guessing this will take some experimenting.

**Review by kimbarrett — COMMENTED** [review]
This is effectively creating a new port.  That shouldn't be done lightly. 

Who needs this?  (Real question, not rhetorical.)  We've discussed things like
this in the past (build for Windows using something other than MSVC), and it
wasn't clear there was a user community sufficient to warrant the effort to
develop and (especially) maintain such a port.

Who is going to maintain this?

Most "secondary" ports have an associated JDK project that is responsible for
them.  This provides a place for other developers to find the maintainers.

Because there was only the one configuration, there have been places that
conflate Windows (the OS) and MSVC (the compiler), both in the build system
and in the source code.  And for source code that means both HotSpot and
native code elsewhere in the JDK.  Julian has been cleaning up some of that,
but I've no idea how much might be left.  And things will quickly bit rot if
there aren't active maintainers.

**Comment by djelinski** [comment]
Thanks @kimbarrett for your input.

I'm not aware of anyone who needs this. I am using clang only for the warnings it reports. It found a number of previously undetected problems.

If new compilers are added on a strictly as-needed basis, this one has no place in the JDK; the Structured Exception Handling implementation in clang is sub-standard, and a hack is required to make it work. The resulting artifacts pass the tests I'm interested in, but I have no intention of running the entire test suite.

I'm only providing this patch to make it easier for others to experiment with clang-cl. If that's not a good enough reason to get this patch merged, I can keep maintaining it in my fork.

**Comment by magicus** [comment]
I tend to agree with both Camp Kim and Camp David/Julian here. In many ways, the code gets healthier the more variation we have, i terms of support for operating systems, architectures, compilers and C libraries. Both the actual code base gets more exposure which can force out bugs that were lurking in the dark, and the build system gets a requirement for clear separations and structure. So in general, I like to see more compilers and platforms in the JDK!

But, then there is Kim's argument about maintenance cost. And it is not to be taken lightly. If the port can be "sneaked in" without much effort or impact on other code, then it essentially free, and even if it is not strictly needed, or we do not have an active maintainer, I see no problem with using it. Great examples of this would be the zero ports on odd CPUs, like MIPS.  The more the port interferes with existing code, the higher this barrier becomes.

But there is also a second cost associated with new ports (I'm counting the new compiler as a port, in this context), and that is the arbitrary differences that do not bring forth any solved bugs, or good abstractions. All these "#ifdef on_this_platform do_it_that_way". If you need many such changes, this also raises the barrier for inclusion.

**Comment by magicus** [comment]
So, in this case, I believe that if it is possible to reuse warnings from "normal" clang, and the rest of the build process from the existing microsoft-cl toolchain, then the impact might be very small in the code. But it will require some surgery in the toolchain logic, to separate the actual compiler from the rest of the toolchain. 

It can certainly be done, but might take a few iterations back and forth to get it right. But maybe that's the sort of thing that @TheShermanTanker enjoys?

So my verdict is that we can't really accept this PR as it is, but using microsoft-clang is definitely an interesting experiment, and we should lay the groundwork for being able to implement it in a proper fashion. And if it then actually can be implemented with minimal impact to existing code, then we should do it.

**Comment by djelinski** [comment]
@magicus just to make sure we are talking about the same thing, are you suggesting configuring the compiler separately from the toolchain, using something like `--with-toolchain=microsoft --with-compiler=clang-cl`?

Technically that could work. It still doesn't address the argument that no organization is going to take up the task of maintaining this setup.

**Comment by magicus** [comment]
Yes, something like that, I guess. That's the best way I see towards modelling the change that occurs when cl.exe is replaced with clang-cl.exe. I have not really thought about the details, but the "toolchain" concept has been a bit loose, basically from day 1. For instance, "clang" might mean "clang on linux" or "clang as part of Xcode", and sometimes we really have meant the latter but expressed it like it included the former also.

We have had a very simple world to model before, where we have had basically 1 os <-> 1 toolchain <-> 1 compiler:
* linux - "gnu toolchain" - gcc
* macOS - Xcode - clang
* windows - visual studio/windows sdk - cl

This has allowed us to be sloppy on exactly what we mean. In some cases we have started to tighten up the distinction between os and compiler+toolchain, but we have not really addressed the situation where the compiler is separate from the rest of the toolchain, even though we have had two such cases (originally, gcc was used alongside clang on macOS, and currently, clang is supported as an alternative to gcc on linux).

So I do not know how we should express it as a command line, or how much change is needed in the code base, but yeah, something like that...

**Comment by djelinski** [comment]
Thanks for the discussion.

**Pull request closed without merging**